File: sdks/ruby/lib/amlexia/tracing.rb
# frozen_string_literal: true

require "securerandom"

module Amlexia
  TraceContext = Struct.new(
    :trace_id, :span_id, :parent_span_id, :session_id, :user_id, :environment, :release_version,
    keyword_init: true
  )

  module Tracing
    module_function

    def create_trace_context(**kwargs)
      TraceContext.new(
        trace_id: kwargs[:trace_id] || SecureRandom.hex(16),
        span_id: kwargs[:span_id] || SecureRandom.hex(8),
        parent_span_id: kwargs[:parent_span_id],
        session_id: kwargs[:session_id],
        user_id: kwargs[:user_id],
        environment: kwargs[:environment],
        release_version: kwargs[:release_version]
      )
    end

    def child_span(parent)
      TraceContext.new(
        trace_id: parent.trace_id,
        span_id: SecureRandom.hex(8),
        parent_span_id: parent.span_id,
        session_id: parent.session_id,
        user_id: parent.user_id,
        environment: parent.environment,
        release_version: parent.release_version
      )
    end
  end
end

File: sdks/python/src/amlexia/tracing.py
"""Distributed tracing helpers for Amlexia Python SDK."""
from __future__ import annotations

import uuid
from dataclasses import dataclass
from typing import Optional


@dataclass
class TraceContext:
    trace_id: str
    span_id: str
    parent_span_id: Optional[str] = None
    session_id: Optional[str] = None
    user_id: Optional[str] = None
    environment: Optional[str] = None
    release_version: Optional[str] = None


def create_trace_context(**kwargs: str) -> TraceContext:
    return TraceContext(
        trace_id=kwargs.get("trace_id", uuid.uuid4().hex),
        span_id=kwargs.get("span_id", uuid.uuid4().hex[:16]),
        parent_span_id=kwargs.get("parent_span_id"),
        session_id=kwargs.get("session_id"),
        user_id=kwargs.get("user_id"),
        environment=kwargs.get("environment"),
        release_version=kwargs.get("release_version"),
    )


def child_span(parent: TraceContext) -> TraceContext:
    return TraceContext(
        trace_id=parent.trace_id,
        span_id=uuid.uuid4().hex[:16],
        parent_span_id=parent.span_id,
        session_id=parent.session_id,
        user_id=parent.user_id,
        environment=parent.environment,
        release_version=parent.release_version,
    )
